### PR TITLE
Add support for administrative levels search parameters on LocationHierarchy endpoint

### PR DESCRIPTION
**Pull request by lincmba**
Resolves https://github.com/onaio/fhir-gateway-extension/issues/64

**Engineer Checklist**

- [ ] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [ ] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [ ] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [ ] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [ ] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java [hunk 1]
```diff
@@ -15,9 +15,15 @@ public class Constants {
...
     public static final String TAG_SEARCH_PARAM = "_tag";
+    public static final String TYPE_SEARCH_PARAM = "type";
+    public static final String DEFAULT_ADMIN_LEVEL_TYPE_URL =
+            "https://smartregister.org/codes/administrative-level";
     public static final String AUTHORIZATION = "Authorization";
     public static final String KEYCLOAK_UUID = "keycloak-uuid";
     public static final String IDENTIFIER = "_id";
+    public static final String MIN_ADMIN_LEVEL = "administrativeLevelMin";
+    public static final String MAX_ADMIN_LEVEL = "administrativeLevelMax";
+    public static final int DEFAULT_MAX_ADMIN_LEVEL = 10;
```
Comment (lincmba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:26]: What is a reasonable default maximum admin level depth?

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java [hunk 2]
```diff
@@ -15,9 +15,15 @@ public class Constants {
     public static final String PARAM_VALUES_SEPARATOR = ",";
     public static final String PROXY_TO_ENV = "PROXY_TO";
     public static final String TAG_SEARCH_PARAM = "_tag";
+    public static final String TYPE_SEARCH_PARAM = "type";
+    public static final String DEFAULT_ADMIN_LEVEL_TYPE_URL =
+            "https://smartregister.org/codes/administrative-level";
```
Comment (lincmba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:20]: Not sure about this url, considering this comment https://github.com/onaio/fhir-web/issues/1349#issuecomment-2033718095. What url is best to go with `.../codes/..` or `.../CodeSystem/...`

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java [hunk 3]
```diff
@@ -11,9 +13,8 @@
 
 import javax.servlet.http.HttpServletRequest;
 
-import org.hl7.fhir.r4.model.Bundle;
-import org.hl7.fhir.r4.model.Identifier;
-import org.hl7.fhir.r4.model.Location;
+import org.hl7.fhir.instance.model.api.IBaseBundle;
+import org.hl7.fhir.r4.model.*;
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java:17]: Avoid wildcard imports

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java [hunk 4]
```diff
@@ -22,6 +23,7 @@
 import org.smartregister.model.location.LocationHierarchy;
 
 import ca.uhn.fhir.rest.client.api.IGenericClient;
+import ca.uhn.fhir.rest.gclient.*;
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java:26]: Same as above

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java (hunk 2 above)
Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:20]: @lincmba let's go with `CodeSystem` since this is what web is using

We will update and document that

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java (hunk 1 above)
Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:26]: I am not sure we can have a `max `level because countries have different hierarchies.  I know we can have the `min` which is `0` but I don't think we can restrict the max 

In case the `max` is not defined we are supposed to fetch the whole hierarchy to the end of the node

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java [hunk 5]
```diff
@@ -107,6 +110,92 @@ public void testGetPaginatedLocationsPaginatesLocations() {
...
+        assertEquals(expectedLevels, adminLevels);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void testGenerateAdminLevelsWithInvalidRange() {
+        locationHierarchyEndpointHelper.generateAdminLevels("3", "1");
+    }
+
+    @Test
+    public void testGetPaginatedLocationsPaginatesLocationsNa() {
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelperTest.java:156]: Is this this testcase name correct? If so then it can be rewritten to be more clear esp. the **Na** suffix

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java (hunk 1 above)
Comment (lincmba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: Seeing this is how the type element is represented:
```
"type": {
    "coding": [
      {
        "system": "http://example.org/fhir/CodeSystem/location-type",
        "code": "1",
        "display": "Level 1"
      }
    ],
```
querying directly with conditions like `code is greater than 5` on a string representation of a code is not possible. FHIR does not support such numeric comparisons on string fields directly within the query. So, to handle this, I have implement post-query filtering in the application logic.

This commit addresses that https://github.com/onaio/fhir-gateway-extension/pull/66/commits/7d9ba8b5154cccb106491d475947e8427ffe4148 
This means, api requests with `adminstrativeLevelMax` set will be more optimized than those without.

Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: This makes sense. Can we review the performance with what we have as we think of how to address this?

Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: How many levels can we have at most? 

How would the performance compare if for example we used an OR operation for the levels since they are numeric? e.g. if we know we have 10 levels, the query for greater than 5 can be where code equals 6 or code equals 7 or code equals 8 e.t.c The assumption is that OR operation works this way.

Comment (lincmba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: @ndegwamartin That was indeed the original implementation, based on the assumption that a maximum depth could be defined. The code constructed an `OR` query for all locations within the specified range, where the maximum value was determined as the greater of the `DEFAULT_MAX_ADMIN_LEVEL` constant and the `administrativeLevelMax` parameter provided by the user, and the minimum value was set to either `0` or the `administrativeLevelMin` parameter.

This functionality will continue to apply in all cases except when a maximum depth cannot be determined. Specifically, this occurs when neither `administrativeLevelMax` nor the `DEFAULT_MAX_ADMIN_LEVEL` constant are defined. As mentioned by @dubdabasoduba, this is a possibility and therefore, in such instances, post-processing will be necessary to filter the data appropriately.

Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: If the performance impact is significant when its open ended I think we are better off making a best guess estimate for a default max. I also think in real life hierarchies cannot go above 10 levels. We could even increase it to 15 or something. 

We can also use data from OpenSRP v1, current OpenSRP v2 and other systems e.g. OpenMRS known support/deployments for what max levels we have seen so far. 

On top of this, we could make that default configurable at runtime incase a deployment needs to change it cc 
 @lincmba @dubdabasoduba

Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java]: @ndegwamartin let's have the default at `DEFAULT_MAX_ADMIN_LEVEL=10` 

Also let's add this to the documentation.

**Review by dubdabasoduba — DISMISSED** [review]

**Review by dubdabasoduba — APPROVED** [review]
